tests/functional/a/arguments_differ.py

```diff
@@ -151,15 +151,24 @@ class SuperClass(object):
     def impl(arg1, arg2, **kwargs):
         return arg1 + arg2
 
+    def should_have_been_decorated_as_static(arg1, arg2):  # pylint: disable=no-self-argument
+        # Enum._generate_next_value in stdlib should be static
+        return arg1 + arg2
+
 
 class MyClass(SuperClass):
 
-    def impl(self, *args, **kwargs):
+    @staticmethod
+    def impl(*args, **kwargs):
         """
         Acceptable use of vararg in subclass because it does not violate LSP.
         """
         super().impl(*args, **kwargs)
 
+    @staticmethod
+    def should_have_been_decorated_as_static(arg1, arg2):

```

Comment:
Imo it should emit argument-differ, there are two errors here, the first one lack a self argument but this one's argument differ as there are 3 instead of 2.

Reply:
Interesting, let me see if I can follow. How are you counting three arguments? I think that's a code-style suggestion to say that there should be no static methods inside classes without a decorator, so if someone _has_ done that, we get here and they just have 2 == 2.

Reply:
Sorry I meant 2 instead of 3. The first function would need self, arg1, arg2 and this one too but as it's static having 2 is equivalent to 3 (like the false positive you found higher up)

Reply:
In the first function ``arg1`` should be called ``self`` but as long as it's not fixed we don't know if it has 2 or 3 args (will the user rename arg1 to self or add self on top of arg1 arg2 ?). Assuming arg1 will be renamed to self and as the second function is static it means it should only have one arg equivalent to ``arg2`` as the self is implicit. If we assume self will be added before arg1 after the user fix it then no error makes more sense. But we have to detect two problems correctly instead of one for that to work, not making assumption is safer imo.

Reply:
> In the first function arg1 should be called self but as long as it's not fixed we don't know if it has 2 or 3 args (will the user rename arg1 to self or add self on top of arg1 arg2 ?).

I guess this is the part I'm having trouble with -- let's say a user has disabled `no-self-use` because they don't care whether their static methods are decorated as such. Now following your idea pylint will be coupling the behavior of `arguments-differ` to the choice of `no-self-use` which will be frustrating for that user. There's nothing strictly wrong with the Enum class in the report, it just makes typing harder. Some users don't type.

> will the user rename arg1 to self or add self on top of arg1 arg2 ?

This is the assumption I would be reluctant to make -- why are we assuming `self` needs to be added at all?

Reply:
> This is the assumption I would be reluctant to make -- why are we assuming self needs to be added at all?

I agree we should not couple ``no-self-argument`` to ``argument-differ``.

But argument differ here:
```python
class Mother:
   def should_have_been_decorated_as_static(arg1, arg2):  # pylint: disable=no-self-argument
        # Enum._generate_next_value in stdlib should be static
        return arg1 + arg2

class Daugther(Mother):
    @staticmethod
    def should_have_been_decorated_as_static(arg1, arg2): #[argument-differ]
        return arg1 + arg2
```
Mother need 1 explicit args (self is implicit + arg2):
```python
>>> m = Mother()
>>> m.should_have_been_decorated_as_static(1, 2 )
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: should_have_been_decorated_as_static() takes 2 positional arguments but 3 were given
>>> m.should_have_been_decorated_as_static(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in should_have_been_decorated_as_static
TypeError: unsupported operand type(s) for +: 'Mother' and 'int'
```

While Daugter needs 2 explicit args:
```python
>>> d = Daugther()
>>> d.should_have_been_decorated_as_static(1, 2)
3
>>> d.should_have_been_decorated_as_static(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: should_have_been_decorated_as_static() missing 1 required positional argument: 'arg2'
```
Thus argument-differ should be triggered.

Reply:
Ah, but only if you call it on the instance!

```python
>>> Mother.should_have_been_decorated_as_static(1, 2)
3
>>> Daugther.should_have_been_decorated_as_static(1, 2)
3
```

I think this was the scenario in the stdlib, which is why I didn't raise a bug against CPython. What do you think?

Reply:
OK it makes sense now. I think ``@staticmethod`` is a stronger signal than the arg not being called self but there is already a message for the argument not being called self.